nighthawk/include/nighthawk/client/benchmark_client.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "envoy/runtime/runtime.h"
+
+#include "nighthawk/common/statistic.h"
+
+namespace Nighthawk {
+namespace Client {
+
+typedef std::function<bool(const std::string, const uint64_t)> CounterFilter;
+
+class BenchmarkClient {
+public:
+  virtual ~BenchmarkClient() = default;
+
+  /**
+   * Initialize will be called on the worker thread after it has started.
+   * @param runtime to be used during initialization.
+   */
+  virtual bool initialize(Envoy::Runtime::Loader& runtime) PURE;
+
+  /**
+   * Terminate will be called on the worker thread before it ends.
+   */
+  virtual void terminate() PURE;
+
+  /**
+   * Turns latency measurement on or off.
+   *
+   * @param measure_latencies true iff latencies should be measured.
+   */
+  virtual void setMeasureLatencies(bool measure_latencies) PURE;
+
+  /**
+   * Gets the statistics, keyed by id.
+   * @return StatisticPtrMap A map of Statistics keyed by id.
+   */
+  virtual StatisticPtrMap statistics() const PURE;
+
+  /**
+   * Tries to start a request. In open-loop mode this MUST always return true.
+   *
+   * @param caller_completion_callback The callback the client must call back upon completion of a
+   * successfully started request.
+   *
+   * @return true if the request could be started.
+   * @return false if the request could not be started, for example due to resource limits.
+   */
+  virtual bool tryStartOne(std::function<void()> caller_completion_callback) PURE;
+
+  /**
+   * Gets a map of tracked counter values, keyed by name.
+   * @param filter function that returns true iff a counter should be included in the map,
+   * based on the named and value it gets passed. The default filter returns all counters.
```

Comment:
FWIW, Envoy also does stats filtering with StatsMatcher.

Reply:
https://github.com/envoyproxy/envoy/blob/4a5f85809cef13be47873483789a5a7bb64ded7c/api/envoy/config/metrics/v2/stats.proto#L76

Reply:
Actually, maybe I misunderstand. Why do we need to filter here? Could we just return the raw stats object and use that?

Reply:
We use the filter to get non-0 valued stats. We can use that to keep the output small, but also in tests to ensure that we check all relevant counters (when we want to).